lib/update/phase_coredns.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	corednsResourceName = "gravity:coredns"
+)
+
+func (r phaseBuilder) corednsPhase(leadMaster storage.Server) *phase {
+	phase := root(phase{
+		ID:          "coredns",
+		Description: "Provision coredns resources",
+		Executor:    coredns,
+		Data: &storage.OperationPhaseData{
+			Server: &leadMaster,
+		},
+	})
+	return &phase
+}
+
+type updatePhaseCoreDNS struct {
+	log.FieldLogger
+	kubernetesOperation
+}
+
+// NewPhaseCoreDNS create an upgrade phase to add coredns rbac permissions
+// HACK: This phase needs to run before updating planet, as coredns is embedded in planet
+// and will come online as planet restarts
+func NewPhaseCoreDNS(c FSMConfig, plan storage.OperationPlan, phase storage.OperationPhase) (*updatePhaseCoreDNS, error) {
+	op, err := newKubernetesOperation(c, plan, phase)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &updatePhaseCoreDNS{
+		kubernetesOperation: *op,
+		FieldLogger: log.WithFields(log.Fields{
+			"phase": phase.ID,
+		}),
+	}, nil
+}
+
+// Execute will add rbac permissions for coredns to sync cluster information
+func (p *updatePhaseCoreDNS) Execute(ctx context.Context) error {
+	_, err := p.kubernetesOperation.Client.RbacV1().ClusterRoles().Create(&rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: corednsResourceName},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Verbs:     []string{"list", "watch"},
+				Resources: []string{"endpoints", "services", "namespaces"},
+			},
+		},
+	})
+	err = rigging.ConvertError(err)
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+
+	_, err = p.kubernetesOperation.Client.RbacV1().ClusterRoleBindings().Create(&rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{Name: corednsResourceName},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     corednsResourceName,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:     "User",
+				Name:     "coredns",
+				APIGroup: "rbac.authorization.k8s.io",
+			},
+		},
+	})
+	err = rigging.ConvertError(err)
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Rollback - Noop (don't worry about deleting resources during a rollback, they'll just be unused)
+func (p *updatePhaseCoreDNS) Rollback(context.Context) error {
+	return nil
+}
+
+func shouldUpdateCoreDNS(client *kubernetes.Clientset) (bool, error) {
+	_, err := client.RbacV1().ClusterRoles().Get(corednsResourceName, metav1.GetOptions{})
+	err = rigging.ConvertError(err)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return true, nil
+		}
+		return false, trace.Wrap(err)
+	}
+
+	_, err = client.RbacV1().ClusterRoleBindings().Get(corednsResourceName, metav1.GetOptions{})
```

Comment:
Is the above check not enough?

Reply:
Yes, only one of the checks itself should be required to see if we're on a pre-coredns version and need to insert the configuration.

I'm just more following the philosophy, of if the cluster appears to be missing configuration that's required for coredns to function, regardless of the reason, the upgrade should detect that and correct the state. Assuming if one object is missing the other will be too is reasonable, it's just this is more likely to work if an unpredictable scenario happens (Something like Partial Upgrade, Rollback, Re-Upgrade scenarios can be difficult to reason about and predict).